src/vm/vm_memory/memory_segments.rs

```diff
@@ -128,6 +136,56 @@ impl MemorySegmentManager {
             Err(MemoryError::WriteArg)
         }
     }
+
+    pub fn get_memory_holes(
+        &self,
+        accessed_addresses: &HashSet<Relocatable>,
+    ) -> Result<usize, MemoryError> {
+        let segment_used_sizes = self
+            .segment_used_sizes
+            .as_ref()
+            .ok_or(MemoryError::MissingSegmentUsedSizes)?;
+
+        let mut accessed_offsets_sets = HashMap::new();
+        for addr in accessed_addresses {
+            if addr.segment_index < 0 {
+                return Err(MemoryError::AddressInTemporarySegment(addr.segment_index));
+            }
+
+            let (index, offset) = from_relocatable_to_indexes(addr);
+            let (segment_size, offset_set) = match accessed_offsets_sets.get_mut(&index) {
+                Some(x) => x,
+                None => {
+                    let segment_size = self
+                        .get_segment_size(index)
+                        .ok_or(MemoryError::SegmentNotFinalized(index))?;
+
+                    accessed_offsets_sets.insert(index, (segment_size, HashSet::new()));
+                    accessed_offsets_sets.get_mut(&index).unwrap()
+                }
+            };
+            if offset > *segment_size {
+                return Err(MemoryError::NumOutOfBounds);
+            }
+
+            offset_set.insert(offset);
+        }
+
+        Ok((0..self.segment_sizes.len())
+            .chain(0..segment_used_sizes.len())
+            .collect::<HashSet<_>>()
+            .into_iter()
+            .map(|index| accessed_offsets_sets.get(&index).unwrap())
+            .map(|(segment_size, offsets_set)| segment_size - offsets_set.len())
+            .sum())
```

Comment:
This is equivalent to:
```suggestion
        let max = cmp::max(self.segment_sizes.len(), segment_used_sizes.len());
        Ok((0..max)
            .map(|index| accessed_offsets_sets.get(&index).unwrap())
            .map(|(segment_size, offsets_len)| segment_size - offsets_set.len())
            .sum())
```
but saving a hash creation and several accesses and reallocations.
I'm also a bit wary of the `unwrap` call. Wouldn't it break if any segment was skipped or created but not used?

Reply:
No excuse for this one :C
I completely missed this by focusing on replicating the functionality from the starknet reference.

Reply:
About the `.unwrap()`, Python doesn't have that problem because it uses the keys from the `accessed_offsets` dictionary where we use the index. Knowing this, I assumed it'd be ok to unwrap here, but now that you mention it...

In our case, if a segment isn't accessed then no value will be created. Since Python would not crash in that case I think modifying it to filter out invalid segments is the right choice.